lib/middleware/process-image-request.js

```
@@ -47,17 +60,75 @@ function processImage(config) {
 			transform.setImmutable(true);
 		}
 
-		// Create a Cloudinary transform
-		const appliedTransform = cloudinaryTransform(transform, {
-			cloudinaryAccountName: config.cloudinaryAccountName,
-			cloudinaryApiKey: config.cloudinaryApiKey,
-			cloudinaryApiSecret: config.cloudinaryApiSecret
-		});
+		let transformReady = Promise.resolve();
+		// 1. get image from destination
+		const originalImageURI = transform.getUri();
+		if (cache[originalImageURI]) {
+			const imageName = cache[originalImageURI];
+			transform.setName(imageName);
+		} else {
+			transformReady = axios.get(encodeURI(originalImageURI), {
+				timeout: 10000, // 10 seconds
+				validateStatus: function (status) {
+					return status >= 200 && status < 600;
+				},
+				responseType: 'arraybuffer'
+			}).then(async imageResponse => {
+				if (imageResponse.status >= 400) {
+					throw httpError(imageResponse.status);
+				}
+				if (imageResponse.headers['content-type'].includes('text/html')) {
+					throw httpError(400);
+				}
+				const file = imageResponse.data;
+
+				// 2. hash image
+				const hash = crypto.createHash('sha256');
+				hash.update(file);
 
-		// Store the transform and applied transform for
-		// use in later middleware
-		request.transform = transform;
-		request.appliedTransform = appliedTransform;
-		next();
+				const imageName = hash.digest('hex');
+				transform.setName(imageName);
+				cache[originalImageURI] = imageName;
+				
+				// 3. check cloudinary to see if image with name exists
+				let imageAlreadyUploadedToCloudinary = false;
+				try {
+					imageAlreadyUploadedToCloudinary = await resource(imageName);
+					// console.log({imageAlreadyUploadedToCloudinary});
+				} catch (imageAlreadyUploadedToCloudinaryError) {
+					// console.dir({imageAlreadyUploadedToCloudinaryError});
```

What happens if this fails silently and we try to upload an image with the same name?

---

Uploading an image with the same name is fine because we have set the upload config `overwrite` to `true`, which tells cloudinary to overwrite the existing image.